pkg/controller/cloneset/cloneset_controller.go

```diff
@@ -536,3 +577,132 @@ func (r *ReconcileCloneSet) claimPods(instance *appsv1alpha1.CloneSet, pods []*v
+	}
+
+	// ignore if this revision can not update in-place
+	opts := &inplaceupdate.UpdateOptions{}
+	inplaceupdate.SetOptionsDefaults(opts)
+	spec := opts.CalculateSpec(currentRevision, updateRevision, opts)
+	if spec == nil {
+		klog.V(4).Infof("CloneSet %s/%s skipped to create ImagePullJob for %s -> %s can not update in-place",
+			cs.Namespace, cs.Name, currentRevision.Name, updateRevision.Name)
+		return r.patchControllerRevisionLabels(updateRevision, appsv1alpha1.ImagePreDownloadIgnoredKey, "true")
+	}
+
+	// start to create jobs
+
+	var pullSecrets []string
+	for _, s := range cs.Spec.Template.Spec.ImagePullSecrets {
+		pullSecrets = append(pullSecrets, s.Name)
+	}
+
+	selector := cs.Spec.Selector.DeepCopy()
+	selector.MatchExpressions = append(selector.MatchExpressions, metav1.LabelSelectorRequirement{
+		Key:      apps.ControllerRevisionHashLabelKey,
+		Operator: metav1.LabelSelectorOpNotIn,
+		Values:   []string{updateRevision.Name, updateRevision.Labels[history.ControllerRevisionHashLabel]},
+	})
+
+	var pullTimeoutSeconds int32 = 300
+	if str, ok := cs.Annotations[appsv1alpha1.ImagePreDownloadTimeoutSecondsKey]; ok {
+		if i, err := strconv.Atoi(str); err == nil {
+			pullTimeoutSeconds = int32(i)
+		}
+	}
+
+	parallelism := intstr.FromInt(1)
+	if str, ok := cs.Annotations[appsv1alpha1.ImagePreDownloadParallelismKey]; ok {
+		if i, err := strconv.Atoi(str); err == nil {
+			parallelism = intstr.FromInt(i)
+		}
+	}
+
+	labelMap := make(map[string]string)
+	for k, v := range cs.Spec.Template.Labels {
+		labelMap[k] = v
+	}
+	labelMap[history.ControllerRevisionHashLabel] = updateRevision.Labels[history.ControllerRevisionHashLabel]
+
+	for name, image := range spec.ContainerImages {
+		job := &appsv1alpha1.ImagePullJob{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:       cs.Namespace,
+				Name:            fmt.Sprintf("%s-%s", updateRevision.Name, name),
+				OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(cs, clonesetutils.ControllerKind)},
+				Labels:          labelMap,
+			},
+			Spec: appsv1alpha1.ImagePullJobSpec{
+				Image:       image,
+				PullSecrets: pullSecrets,
+				PodSelector: &appsv1alpha1.ImagePullJobPodSelector{LabelSelector: *selector},
+				Parallelism: &parallelism,
+				PullPolicy:  &appsv1alpha1.PullPolicy{BackoffLimit: utilpointer.Int32Ptr(1), TimeoutSeconds: &pullTimeoutSeconds},
+				CompletionPolicy: appsv1alpha1.CompletionPolicy{
+					Type:                    appsv1alpha1.Always,
+					TTLSecondsAfterFinished: utilpointer.Int32Ptr(600),
+				},
+			},
+		}
+
+		err := r.Create(context.TODO(), job)
+		if err != nil {
+			if !errors.IsAlreadyExists(err) {
+				klog.Errorf("CloneSet %s/%s failed to create ImagePullJob %s: %v", cs.Namespace, cs.Name, job.Name, err)
+			}
+			continue
+		}
+		klog.V(3).Infof("CloneSet %s/%s created ImagePullJob %s for image: %s", cs.Namespace, cs.Name, job.Name, image)
```

Comment:
shall we emit an event as cloneset about the imagepulljobs here ?

Reply:
done